File: java/src/eti/Project2/Organisms/Hogweed.java
package eti.Project2.Organisms;

import eti.Project2.GUI;
import eti.Project2.MapPosition;
import eti.Project2.World;

/**
 * Created By IDEA
 * Author: s188026 Yifei Liu
 * Date: 2021/5/30 14:12
 * Description: Hogweed     '%':
 *      Kills every animal in its immediate neighborhood except cyber-sheep.
 *      Kills any animal which eats it, apart from cyber-sheep.
 */

public class Hogweed extends Plant{
    //constructor
    public Hogweed(char n, int s, int i, int X, int Y, World w) {
        super(n, s, i, X, Y, w);
    }

    public Hogweed(Hogweed h) {
        super(h);
    }

    public Hogweed(int X, int Y, World world) {
        setName('%');
        setStrength(10);
        setInitiative(0);
        setPosition(new MapPosition(X, Y));
        setWorld(world);
    }

    @Override
    public void Action()
    {
        MapPosition defenderP = new MapPosition(-1, -1);

        /**std::cout + "'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") is trying to killing animal around" + std::endl;*/
        GUI.getInformation().append("'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") is killing animal around\n");
        
        //kill WASD animal around(so it didn't kill the plant)
        defenderP = getPosition().Up(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
            getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
            getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
            getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
            getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
            getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().Left(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().Down(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().Right(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().LeftUp(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().LeftDown(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().RightUp(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }

        defenderP = getPosition().RightDown(getWorld());
        if (getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'w' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 's' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'f' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 't' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'a' ||
                getWorld().organisms[defenderP.getY()][defenderP.getX()].getName() == 'h')
        {
            getWorld().organisms[defenderP.getY()][defenderP.getX()] = new Organism('_', -1, -1, defenderP.getX(), defenderP.getY(), getWorld());
        }
    }

    @Override
    public void Collision(Organism attacker)
    {
        if (attacker.getName() != 'c')
        {
            //who eats this will die(so this will be eaten)
            /**std::cout + "'" + getName() + "' kills attacker who eats it'" + (Grid(attacker.getPosition().getX(), attacker.getPosition().getY())).getName() + "'(" + attacker.getPosition().getX() + "," + attacker.getPosition().getY() + ")" + std::endl;*/
            GUI.getInformation().append( "'" + getName() + "' kills attacker who eats it'" + attacker.getName() + "'(" + attacker.getPosition().getX() + "," + attacker.getPosition().getY() + ")\n");

            getWorld().organisms[getPosition().getY()][getPosition().getX()] = new Organism('_', -1, -1, getPosition().getX(), getPosition().getY(), getWorld());
        }
        else
        {
            /**std::cout + "'" + getName() + "' is eaten by cybersheep" + std::endl;*/
            GUI.getInformation().append("'" + getName() + "' is eaten by cyber sheep\n");

            attacker.setPosition(this.getPosition());
            attacker.setPosition(this.getPosition());
            getWorld().organisms[getPosition().getY()][getPosition().getX()] = attacker;
        }
    }
}

File: cpp/Hogweed.cpp
#include "Hogweed.h"
#include <iostream>

Hogweed::Hogweed(int X, int Y, World* world) :Plant(HOGWEED, 10, 0, X, Y, world) {};

void Hogweed::Action()
{
	Position defenderP(-1, -1);
	
	std::cout << "'" << GetName() << "'(" << GetPosition().X << "," << GetPosition().Y << ") is killing animal around" << std::endl;

	//kill WASD animal around(so it didn't kill the plant)
	defenderP = GetPosition().Up(GetWorld());
	if ((Grid(defenderP.X, defenderP.Y))->GetName() == WOLF ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == SHEEP ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == FOX ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == TURTLE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == ANTELOPE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == HUMAN)
	{
		
		Organism* temp = Grid(defenderP.X, defenderP.Y);
		Grid(defenderP.X, defenderP.Y) = new nullGrid(defenderP);
		delete(temp);
	}

	defenderP = GetPosition().Left(GetWorld());
	if ((Grid(defenderP.X, defenderP.Y))->GetName() == WOLF ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == SHEEP ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == FOX ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == TURTLE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == ANTELOPE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == HUMAN)
	{
		Organism* temp = Grid(defenderP.X, defenderP.Y);
		Grid(defenderP.X, defenderP.Y) = new nullGrid(defenderP);
		delete(temp);
	}

	defenderP = GetPosition().Down(GetWorld());
	if ((Grid(defenderP.X, defenderP.Y))->GetName() == WOLF ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == SHEEP ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == FOX ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == TURTLE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == ANTELOPE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == HUMAN)
	{
		Organism* temp = Grid(defenderP.X, defenderP.Y);
		Grid(defenderP.X, defenderP.Y) = new nullGrid(defenderP);
		delete(temp);
	}

	defenderP = GetPosition().Right(GetWorld());
	if ((Grid(defenderP.X, defenderP.Y))->GetName() == WOLF ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == SHEEP ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == FOX ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == TURTLE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == ANTELOPE ||
		(Grid(defenderP.X, defenderP.Y))->GetName() == HUMAN)
	{
		Organism* temp = Grid(defenderP.X, defenderP.Y);
		Grid(defenderP.X, defenderP.Y) = new nullGrid(defenderP);
		delete(temp);;
	}
}

void Hogweed::Collision(Organism*& attacker)
{
	/*if (attacker->GetPosition().X != -1)
		Grid(attacker->GetPosition().X, attacker->GetPosition().Y) = new nullGrid(attacker->GetPosition());*/
	if (attacker->GetName() != CYBERSHEEP)
	{
		//who eats this will die(so this will be eaten)
		std::cout << "'" << GetName() << "' kills attacker who eats it'" << (Grid(attacker->GetPosition().X, attacker->GetPosition().Y))->GetName() << "'(" << attacker->GetPosition().X << "," << attacker->GetPosition().Y << ")" << std::endl;

		delete(attacker);

		Grid(GetPosition().X, GetPosition().Y) = new nullGrid(GetPosition());
	}
	else
	{
		std::cout << "'" << GetName() << "' is eaten by cybersheep" << std::endl;

		(Grid(GetPosition().X, GetPosition().Y)) = attacker;
		attacker->SetPosition(this->GetPosition());
		attacker->SetPosition(this->GetPosition());
	}

	delete(this);
}